File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int lfx[n+1], rfx[n+1];
        lfx[0] = 0;
        for(int i = 0; i < n; i++)
        {
            lfx[i+1] = max(lfx[i], height[i]);
        }
        rfx[n] = 0;
        for(int i = n-1; i >= 0; i--)
        {
            rfx[i] = max(rfx[i+1], height[i]);
        }
        int res = 0;
        for(int i = 0; i < n; i++)
        {
            res += max(min(lfx[i], rfx[i+1])-height[i], 0);
        }
        return res;
    }
};

File: Data Structures & Algorithms/swim-in-rising-water/submission-0.py
import heapq

class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        pq = []
        heapq.heappush(pq, (grid[0][0], 0, 0))
        grid[0][0] = -1
        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]
        n = len(grid)
        res = 0
        while len(pq) > 0:
            height, x, y = heapq.heappop(pq)
            res = max(height, res)
            if x == (n - 1) and y == (n - 1):
                return res
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if nx < 0 or ny < 0 or nx >= n or ny >= n or grid[nx][ny] == -1:
                    continue
                heapq.heappush(pq, (grid[nx][ny], nx, ny))
                grid[nx][ny] = -1
        
